Convert a lot more components to functional style, and other cleanups

sfc-gh-tteixeira opened the pull request:
## Describe your changes

- Convert many components to functional
    - Components that were straight applications of `useBasicWidgetState`: 
      - `st.time_input`
      - `st.date_input`
      - `st.color_picker`
      - `st.selectbox`
  - Components that require an extra "commit" step to save values to the state:
      - `st.text_input`
      - `st.text_area`
  - Components that have a "commit" step and other complexities:
      - `st.slider` / `st.select_slider`
- Rename `setValueWSource` to `setValueWithSource`
- Move some usage of `FormClearHelper` to `useFormClearHelper`: `st.number_input`, `st.dataframe`, `st.button_group`
- Clean up `st.number_input` a little
- Change `st.slider` algorithm

## GitHub Issue Link (if applicable)

## Testing Plan

- ~~Explanation of why no additional tests are needed~~
- [x] Unit Tests (JS and/or Python)
- [x] E2E Tests
- Any manual testing needed?
  - Yes! Would love tests of:
    - All widgets above + session state
    - All widgets above + developer changes some parameters in the source code   

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-tteixeira committed:
Checkbox checkpoint. Still broken.

sfc-gh-tteixeira committed:
Update Checkbox widget to functional component.

sfc-gh-tteixeira committed:
Introduce useBasicWidgetState and convert Checkbox and Radio to it.

sfc-gh-tteixeira committed:
Make useFormClearHelper follow latest pattern

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into checkbox

sfc-gh-tteixeira committed:
Switching some widgets to use FormClearHelper (need to fix NumberInput still)

sfc-gh-tteixeira committed:
Fix tests that required async/await

sfc-gh-tteixeira committed:
Address code review comments, lint, etc.

sfc-gh-tteixeira committed:
Forgot to rename props type

sfc-gh-tteixeira committed:
Make useFormClearHelper() use WidgetStateManager directly rather than FormClearHelper.

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into checkbox

sfc-gh-tteixeira committed:
Merge branch 'checkbox' into formcleaner

sfc-gh-tteixeira committed:
Also convert underlying /shared/Radio widget to a function component.

sfc-gh-tteixeira committed:
Merge branch 'checkbox' into formcleaner

sfc-gh-tteixeira committed:
Fix radio test

sfc-gh-tteixeira committed:
Fix NumberInput

sfc-gh-tteixeira committed:
Merge branch 'checkbox' into formcleaner

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into formcleaner

sfc-gh-tteixeira committed:
Fix types for useBasicWidgetState

sfc-gh-tteixeira committed:
Make TimeInput functional

sfc-gh-tteixeira committed:
Make DateInput functional

sfc-gh-tteixeira committed:
Make ColorPicker functional

sfc-gh-tteixeira committed:
Make Selectbox functional

sfc-gh-tteixeira committed:
Add tests

sfc-gh-tteixeira committed:
Make Slider functional

sfc-gh-tteixeira committed:
Make TextInput functional

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into formcleaner

sfc-gh-tteixeira committed:
Fix ColorPicker tests

sfc-gh-tteixeira committed:
Remove unused state interface

sfc-gh-tteixeira committed:
Make TextArea functional

sfc-gh-tteixeira committed:
Fix slider focus state

sfc-gh-tteixeira committed:
Fix tests

sfc-gh-tteixeira committed:
Rename ValueWSource to ValueWithSource

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into formcleaner

sfc-gh-tteixeira committed:
Lint lint lint lint lint

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into functionalcomponents

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -282,159 +149,294 @@ class Slider extends React.PureComponent<Props, State> {
       }
     }
-  }
+  })
```
the deps array is missing for the `useEffect` hook which means it will run on every re-render. Is this intended? Otherwise, it looks like `[thumbRefs, thumbValueRefs]` should be passed.

sfc-gh-tteixeira replied:
Yes this is intended. This was previously run on every render as well, so no change here.

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -282,159 +149,294 @@ class Slider extends React.PureComponent<Props, State> {
+      setUiValue(value)
+      debouncedSetValueWithSource(value)
+    },
+    [setValueWithSource]
```
the callback does not use `setValueWithSource`. Did you mean to include `debouncedSetValueWithSource`?

Also, according to the [React doc](https://legacy.reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often), the 
> (The identity of the setCount function is guaranteed to be stable so it’s safe to omit.)

where `setCount` is their useState hooks setter in the example, so its the equivalent to the `setValueWithSource` here. Our linter also does not complain when it is missing from the deps array, which is why I lean towards leaving all `set*` hook-setters out of all the dependency arrays to have a consistent pattern in the codebase 🙂

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -282,159 +149,294 @@ class Slider extends React.PureComponent<Props, State> {
-      )
+  // When resetting a form, `value` will change so we need to change `uiValue`
+  // to match.
+  useEffect(() => {
+    if (value !== uiValue) {
+      setUiValue(value)
     }
+    // Don't include `uiValue` in the deps below or the slider will become
+    // jittery.
+    /* eslint-disable react-hooks/exhaustive-deps */
+  }, [value])
```
According to the [React docs](https://react.dev/reference/react/useState#setstate-caveats) setting the state to the same value does not trigger a re-render:
> If the new value you provide is identical to the current state, as determined by an [Object.is](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) comparison, React will skip re-rendering the component and its children. This is an optimization. Although in some cases React may still need to call your component before skipping the children, it shouldn’t affect your code.

Based on this, I would remove the `if (value !== uiValue)` check and, thus, remove disabling eslint 🙂 

```typescript
  useEffect(() => {
    setUiValue(value)
  }, [value])
```

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -282,159 +149,294 @@ class Slider extends React.PureComponent<Props, State> {
+  }, [])
+
+  const renderThumb = useCallback(
+    /* eslint-disable react/display-name */
```
nit: we can get rid of disabling eslint here by doing this:

```typescript
    React.forwardRef<HTMLDivElement, StyleProps>(function renderThumb(
      props: StyleProps,
      ref
    ): ReactElement {
      const { $thumbIndex } = props
// keep rest like it is
```

I think the advantage is that upon errors we get a better trace with a defined name and we don't disable eslint 😁

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -282,159 +149,294 @@ class Slider extends React.PureComponent<Props, State> {
-      element.formId,
-      this.onFormCleared
-    )
+    // Only run this on first render. No real reason, just nicer.
+    /* eslint-disable react-hooks/exhaustive-deps */
+  }, [])
```
If `disabled` for example changes dynamically in the code, could this lead to an error here because the component is not updated? Do you see an issue with passing `[formattedMinValue, formattedMaxValue, disabled]`?

raethlein replied:
Just tested it. If you have code like this:

```python
st.slider("Label 4", 10000, 25000, 10000, disabled=st.checkbox("Disable"))
```

and you check the disable box, the tickBars are not affected live and stay enabled / disabled. So we should pass `[formattedMinValue, formattedMaxValue, disabled]` here to make it work. 
Perhaps we should update our `e2e_playwright/st_slider.py` test with this^ interactivity to catch this automatically (or at least we could think about filing a tech debt item; I have the feeling that our tests do not go into much activity details like this - and maybe it would be too much overhead to test any single thing in the tests)

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -282,159 +149,294 @@ class Slider extends React.PureComponent<Props, State> {
+      }
+    ),
+    // Only run this on first render, to avoid losing the focus state.
+    /* eslint-disable react-hooks/exhaustive-deps */
+    []
```
Likely the same situation here as mentioned in the other comment about components not updating live

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -282,159 +149,294 @@ class Slider extends React.PureComponent<Props, State> {
+  thumb: HTMLDivElement | null,
+  thumbValue: HTMLDivElement | null
+): void {
+  if (slider && thumb && thumbValue) {
```
super-nit (personal style): I personally prefer writing blocks that encapsulate many lines like this:

```typescript
function alignValueOnThumb(
  slider: HTMLDivElement | null,
  thumb: HTMLDivElement | null,
  thumbValue: HTMLDivElement | null
): void {
  if (!slider || !thumb || !thumbValue) {
    return
  } 

  const ...
} 
```

the reason is that we have one less level of nesting which makes code usually more pleasant and easier to read (its more so if the body itself contains additional nesting).

sfc-gh-tteixeira replied:
Agreed! I was just trying not to modify the existing code too much, to avoid introducing bugs.

raethlein commented on frontend/lib/src/components/widgets/DateInput/DateInput.tsx:
```diff
@@ -74,361 +59,312 @@ function datesToStrings(dates: Date[]): string[] {
+
+  // We need to convert the provided format into a mask supported by the Baseweb datepicker
+  // Thereby, we need to replace all letters with 9s which refers to any number.
+  const dateMask = element.format.replaceAll(/[a-zA-Z]/g, "9")
+  // The Baseweb datepicker supports the date-fns notation for date formatting which is
+  // slightly different from the momentJS notation. Therefore, we need to
+  // convert the provided format into the date-fns notation:
+  const dateFormat = element.format.replaceAll("Y", "y").replaceAll("D", "d")
```
nit: I could imagine that `replaceAll` with a regex is on the more expensive side of operations; should we wrap it in a `useMemo` call or so?

raethlein commented on frontend/lib/src/components/widgets/NumberInput/NumberInput.tsx:
```diff
@@ -263,38 +268,35 @@ export const NumberInput: React.FC<Props> = ({
     commitValue({ value: value ?? null, source: { fromUi: false } })
-  }
+    /* eslint-disable react-hooks/exhaustive-deps */
+  }, [element, step])
```
since we use `useCallback` for `commitValue`, it should be a stable reference when nothing changes and so we should be able to pass it to the deps array and get rid of the eslint-disablement. Maybe I am missing something though.

raethlein commented on frontend/lib/src/components/widgets/NumberInput/NumberInput.tsx:
```diff
@@ -263,38 +268,35 @@ export const NumberInput: React.FC<Props> = ({
     setIsFocused(false)
-  }
+    /* eslint-disable react-hooks/exhaustive-deps */
+  }, [dirty])
```
if we don't pass `value` as a dependency, it might be possible that we commit an old value. Does anything speak against including this and `commitValue` (which is a stable reference because of `useCallback`)?

raethlein commented on frontend/lib/src/components/widgets/Checkbox/Checkbox.tsx:
```diff
@@ -54,23 +54,24 @@ function Checkbox({
+      setValueWithSource({ value: e.target.checked, fromUi: true })
     },
-    [setValueWSource]
+    [setValueWithSource]
```
nit: see [this comment](https://github.com/streamlit/streamlit/pull/9494#discussion_r1764650639) about potentially leaving this dependency out

sfc-gh-tteixeira replied:
I'd love to remove that dependency, but ESLint complains. I think it doesn't simply looks at the `set*` prefix, but actually checks whether the value came from `useState`/`Callaback`/`Memo`/etc.

```
frontend/lib/src/components/widgets/Checkbox/Checkbox.tsx
  75:5  warning  React Hook useCallback has a missing dependency: 'setValueWithSource'. Either include it or remove the dependency array  react-hooks/exhaustive-deps
```

raethlein replied:
ohh interesting! thanks for double-checking 👍 in this case we could also probably remove the comment again as that is the default behavior and I just did not know 😬

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into functionalcomponents

sfc-gh-tteixeira committed:
Address some comments and lint

sfc-gh-tteixeira committed:
Remove async/await in act() when not necessary.

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into functionalcomponents

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into functionalcomponents

sfc-gh-tteixeira committed:
Fix slider value positioning algorithm.

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into functionalcomponents

sfc-gh-tteixeira commented:
@raethlein , finally ready for another review! 

* I merged the latest `develop`, which actually included a lot of the changes from my original PR (yay!)
* I addressed your comments
* In the process of addressing your comments I found some issues with how Sliders draw their values, so I spent some time fixing it. This should really be a separate PR, but I'm being lazy. LMK if you'd like me to split it off!

sfc-gh-tteixeira committed:
Merge

sfc-gh-tteixeira committed:
Lint

raethlein commented on frontend/lib/src/components/widgets/DateInput/DateInput.tsx:
```diff
@@ -74,360 +65,320 @@ function datesToStrings(dates: Date[]): string[] {
+
+  const [isEmpty, setIsEmpty] = useState(false)
+
+  const theme = useTheme()
+  const { colors, fontSizes, lineHeights, spacing } = theme
```
super nit: we could unpack it in one line: `const { colors, fontSizes, lineHeights, spacing } = useTheme()` and also add `sizes` here and do not re-use `theme.spacing`/`theme.sizes` further below and thus do not mix it with the destructured access

raethlein commented on frontend/lib/src/components/widgets/DateInput/DateInput.tsx:
```diff
@@ -74,360 +65,320 @@ function datesToStrings(dates: Date[]): string[] {
+          CalendarContainer: {
+            style: {
+              fontSize: fontSizes.sm,
+              paddingRight: theme.spacing.sm,
+              paddingLeft: theme.spacing.sm,
+              paddingBottom: theme.spacing.sm,
+              paddingTop: theme.spacing.sm,
```
nit: use the destructured property from [here](https://github.com/streamlit/streamlit/pull/9494/files#r1808111702) so that we do not mix the different access ways

raethlein commented on frontend/lib/src/components/widgets/DateInput/DateInput.tsx:
```diff
@@ -74,360 +65,320 @@ function datesToStrings(dates: Date[]): string[] {
+    }: {
+      date: Date | (Date | null | undefined)[] | null | undefined
+    }): void => {
+      if (date === null || date === undefined) {
```
nit (codebase style): we have a helper function `isNullOrUndefined` for this

raethlein commented on frontend/lib/src/components/widgets/DateInput/DateInput.tsx:
```diff
@@ -74,360 +65,320 @@ function datesToStrings(dates: Date[]): string[] {
+  }, [isEmpty, element, setValueWithSource])
+
+  return (
+    <div className="stDateInput" data-testid="stDateInput" style={style}>
```
instead of setting `style={style}` we should probably use a `styled-components.ts` file here and define the `div` with the styling there, because we usually avoid inline styling

Update: oh I have just realized that it has been like this also in the old code, so feel free to disregard this

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -52,249 +58,104 @@ const DEBOUNCE_TIME_MS = 200
+    fragmentId,
+  })
+
+  // We tie the UI to `uiValue` rather than `value` becase `value` only updates
```
nit: `becase` -> `because`

raethlein replied:
```suggestion
  // We tie the UI to `uiValue` rather than `value` because `value` only updates
```

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -308,133 +169,408 @@ class Slider extends React.PureComponent<Props, State> {
-      widgetMgr,
-      element.formId,
-      this.onFormCleared
+    // If, after rendering, if the thumb value's is outside the container (too
```
nit: `If, after rendering, if the` -> ` If, after rendering, the`

raethlein replied:
```suggestion
    // If, after rendering, the thumb value's is outside the container (too
```

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -308,133 +169,408 @@ class Slider extends React.PureComponent<Props, State> {
+  const thumbsAreFarApart =
+    thumb2ValueInnerEdge - thumb1ValueInnerEdge > labelGap
+
+  // If thumbs are far apart, just handle each separately.
```
❤️ awesome comments, thanks!

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -308,133 +169,408 @@ class Slider extends React.PureComponent<Props, State> {
+
+  // 4. If one of the thumbs is so close to the edge that doing the outward
+  // alignment from #3 would cause its value to overflow past the edge, then
+  // try centering the value. And plece the other thumb's value right next to
```
```suggestion
  // try centering the value. And place the other thumb's value right next to
```

raethlein commented on frontend/lib/src/components/widgets/Slider/Slider.tsx:
```diff
@@ -308,133 +169,408 @@ class Slider extends React.PureComponent<Props, State> {
+ * - Keep thumb values within the bounds of the slider.
+ * - Avoid visual jank while moving the thumbs
+ */
+function fixLabelOverlap(
```
Do you think its possible to add some tests for this function since it is not completely trivial? Perhaps either unit tests or maybe a few snapshot e2e playwright tests with a couple of these scenarios. Otherwise I am a little bit worried that no one will ever dare to touch the function again in the future (despite the great comments) or changing anything would involve some manual testing to have confidence that nothing broke

raethlein reviewed (approved):
LGTM! Great that more and more widgets are modernized, thank you 🚀 
I love the slider update and am totally fine to do it in one PR; the only ask I have is to add some form of testing (see [my comment](https://github.com/streamlit/streamlit/pull/9494/files#r1808130526)). Since adding some tests can be a little bit more time involved, maybe we can have it as a separate PR and find someone from the team to add them if you are busy right now 🙂 What do you think?
Other than that I have just left a few nits and feel free to merge after you had a look!

sfc-gh-tteixeira committed:
Update frontend/lib/src/components/widgets/Slider/Slider.tsx

Co-authored-by: Benjamin Räthlein <benjamin.raethlein@gmail.com>

sfc-gh-tteixeira committed:
Update frontend/lib/src/components/widgets/Slider/Slider.tsx

Co-authored-by: Benjamin Räthlein <benjamin.raethlein@gmail.com>

sfc-gh-tteixeira committed:
Update frontend/lib/src/components/widgets/Slider/Slider.tsx

Co-authored-by: Benjamin Räthlein <benjamin.raethlein@gmail.com>

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into functionalcomponents

sfc-gh-tteixeira committed:
Add tests for slider overlaps

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/functionalcomponents' into functionalcomponents

sfc-gh-tteixeira commented:
I added screenshot tests for the slider overlap algorithm. I still need to update the screenshots, but I'm waiting for the automation to do it for me.

sfc-gh-tteixeira committed:
Address comments and fix test

raethlein committed:
Merge branch 'develop' into functionalcomponents

raethlein committed:
Fix import issue

raethlein committed:
Update snapshots
